org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/codeassist/DOMCompletionEngine.java

```diff
@@ -574,6 +612,35 @@ public void complete(org.eclipse.jdt.internal.compiler.env.ICompilationUnit sour
 					if (!fieldAccessType.isRecovered()) {
 						processMembers(fieldAccess, fieldAccessExpr.resolveTypeBinding(), specificCompletionBindings, false);
 						publishFromScope(specificCompletionBindings);
+						if (this.extendsOrImplementsInfo == null && (!this.requestor.isIgnored(CompletionProposal.METHOD_REF_WITH_CASTED_RECEIVER)
+								|| !this.requestor.isIgnored(CompletionProposal.FIELD_REF_WITH_CASTED_RECEIVER))) {
+							
+							IVariableBinding variableToCast = null;
+							if (fieldAccessExpr instanceof Name name && name.resolveBinding() instanceof IVariableBinding variableBinding) {
+								variableToCast = variableBinding;
+							} else if (fieldAccessExpr instanceof FieldAccess parentFieldAccess && parentFieldAccess.resolveFieldBinding() != null) {
+								variableToCast = parentFieldAccess.resolveFieldBinding();
+							}
+							
+							if (variableToCast != null) {
+								Set<String> alreadySuggestedBindingKeys = specificCompletionBindings.all()
+										.map(IBinding::getKey).collect(Collectors.toSet());
+								List<ITypeBinding> castedTypes = findInstanceofChecks(this.toComplete, variableToCast);
+								for (ITypeBinding castedType : castedTypes) {
+									Bindings castedBindings = new Bindings();
+									processMembers(this.toComplete, castedType, castedBindings, false);
+									castedBindings.all().forEach(castedBinding -> {
+										if (alreadySuggestedBindingKeys.contains(castedBinding.getKey())) {
+											return;
+										}
+										if (!this.pattern.matchesName(this.prefix.toCharArray(), castedBinding.getName().toCharArray())) {
+											return;
+										}
+										this.requestor.accept(toCastedReceiverProposal(castedBinding, castedType, fieldAccessExpr));
+									});
+								}
+							}
+						}
```

Comment:
This logic is essentially duplicated in the qualified name section,
I can probly extract a function to avoid reuse